arc/src/ArcToken.sol

```diff
@@ -0,0 +1,486 @@
+    /**
+     * @dev Toggles transfer restrictions. When `transfersAllowed` is true, anyone can transfer tokens.
+     * When false, only whitelisted addresses can send/receive tokens.
+     */
+    function setTransfersAllowed(
+        bool allowed
+    ) external onlyRole(ADMIN_ROLE) {
+        _getArcTokenStorage().transfersAllowed = allowed;
+        emit TransfersRestrictionToggled(allowed);
+    }
+
+    /**
+     * @dev Returns true if token transfers are currently unrestricted (open to all).
+     */
+    function transfersAllowed() external view returns (bool) {
+        return _getArcTokenStorage().transfersAllowed;
+    }
+
+    // -------------- Minting and Burning --------------
+    /**
+     * @dev Mints new tokens to an account. Only accounts with MANAGER_ROLE can call this.
+     * The recipient must be whitelisted if transfers are restricted.
+     */
+    function mint(address to, uint256 amount) external onlyRole(MANAGER_ROLE) {
+        _mint(to, amount);
+    }
+
+    /**
+     * @dev Burns tokens from an account, reducing the total supply. Only accounts with MANAGER_ROLE can call this.
+     */
+    function burn(address from, uint256 amount) external onlyRole(MANAGER_ROLE) {
+        _burn(from, amount);
+    }
+
+    // -------------- Yield Distribution --------------
+    /**
+     * @dev Sets or updates the ERC20 token to use for yield distribution (e.g., USDC).
+     * Only accounts with YIELD_MANAGER_ROLE can update this.
+     */
+    function setYieldToken(
+        address yieldTokenAddr
+    ) external onlyRole(YIELD_MANAGER_ROLE) {
+        if (yieldTokenAddr == address(0)) {
+            revert InvalidYieldTokenAddress();
+        }
+        _getArcTokenStorage().yieldToken = yieldTokenAddr;
+        emit YieldTokenUpdated(yieldTokenAddr);
+    }
+
+    /**
+     * @dev Get a preview of the yield distribution for token holders.
+     * This allows the yield distributor to check how much each holder would receive before
+     * actually distributing yield.
+     * @param amount The amount of yield token to preview distribution for
+     * @return holders Array of token holder addresses
+     * @return amounts Array of amounts each holder would receive
+     */
+    function previewYieldDistribution(
+        uint256 amount
+    ) external view returns (address[] memory holders, uint256[] memory amounts) {
+        ArcTokenStorage storage $ = _getArcTokenStorage();
+        if ($.yieldToken == address(0)) {
+            revert YieldTokenNotSet();
+        }
+        if (totalSupply() == 0) {
+            revert NoTokensInCirculation();
+        }
+
+        uint256 holderCount = $.holders.length();
+        holders = new address[](holderCount);
+        amounts = new uint256[](holderCount);
+
+        uint256 supply = totalSupply();
+        uint256 totalPreviewAmount = 0;
+
+        // Calculate distribution for all but last holder
+        for (uint256 i = 0; i < holderCount; i++) {
```

Comment:
Can you check how many loops we can do under 30m gas? And use that as the maxHolderCount